File: scenarios/resilience.ts
#!/usr/bin/env bun
// scenarios/resilience.ts
//
// Proves Track C (resilience): when the remote host goes away — modeled here
// by `docker stop` of the container hosting helmor-server — Helmor's liveness
// loop notices, flips the runtime to Degraded → Disconnected, surfaces a
// banner across the top of the app, and offers a one-click Reconnect that
// re-establishes SSH + the JSON-RPC pipe once the host comes back.
//
// This is the headline "is this production-ready?" video: a real failure
// (container down), a real detection (liveness ping fails), a real recovery
// (Reconnect mutation succeeds, runtime goes green again).
//
// Preconditions: `bun run dev` (debug build → bridge), helmor-test-linux-arm64
// container running + a connected `docker-linux-arm64` runtime. The scenario
// stops the container, waits for state to flip, restarts it, and reconnects.

import { type Assertion, Tape } from "./lib.ts";

const NAME = process.env.RUNTIME_NAME ?? "docker-linux-arm64";
const CONTAINER = process.env.CONTAINER ?? "helmor-test-linux-arm64";
const OUT = process.env.TAPE_DIR ?? "./tapes/resilience";

const tape = new Tape("resilience", OUT);
await tape.connect();

async function docker(...args: string[]): Promise<string> {
	const p = Bun.spawn(["docker", ...args], { stdout: "pipe", stderr: "pipe" });
	const code = await p.exited;
	const out = await new Response(p.stdout).text();
	if (code !== 0) throw new Error(`docker ${args.join(" ")} → ${code}: ${await new Response(p.stderr).text()}`);
	return out.trim();
}

async function waitForState(target: (s: string) => boolean, timeoutMs: number): Promise<string> {
	const deadline = Date.now() + timeoutMs;
	let last = "(unknown)";
	while (Date.now() < deadline) {
		const rts = (await tape.invoke("list_remote_runtimes", {})) as Array<{
			name: string;
			state?: { type?: string };
		}>;
		const r = rts.find((x) => x.name === NAME);
		last = r?.state?.type ?? "(missing)";
		if (target(last)) return last;
		await tape.sleep(500);
	}
	return last;
}

// 0. Make sure we start connected — if the user's been kicking the runtime,
//    reconnect first so the failure→recovery story is meaningful.
{
	const initialState = await waitForState((s) => s === "connected", 5_000);
	if (initialState !== "connected") {
		tape.log(`runtime ${NAME} state=${initialState}; reconnecting first`);
		await tape.invoke("reconnect_remote_runtime", { name: NAME }, 60_000);
		await waitForState((s) => s === "connected", 30_000);
	}
}

// Reload to clean shell, then open Remote Servers so the row is visible too.
await tape.js('window.location.reload(); return "r";');
await tape.sleep(6000);
await tape.openSettings("remote-servers");
tape.assert("panel_opens", await tape.waitFor("[role=dialog]"));
tape.assert("row_present", await tape.waitFor(`[data-testid=remote-server-row-${NAME}]`, 10_000));

// ── Scene 1: connected baseline ─────────────────────────────────────
await tape.scene({
	caption: `Baseline: ${NAME} is connected — helmor-server is alive on the container`,
	hold: 4,
});

// ── Scene 2: kill the host. Close the dialog so the banner is visible. ─
await tape.closeDialog();
await tape.sleep(400);
tape.log(`stopping container ${CONTAINER}`);
await docker("stop", "-t", "1", CONTAINER);

// Liveness pings every 5s with a 3s timeout. The state flip is usually
// visible within one cycle; wait up to 20s for it to settle on a non-
// connected state.
const downState = await waitForState((s) => s !== "connected", 20_000);
tape.assert("state_flips_offline", downState !== "connected", `state=${downState}`);
tape.assert(
	"banner_appears",
	await tape.waitFor(`[data-testid=remote-connection-banner-row-${NAME}]`, 10_000),
	"top-of-shell banner",
);

await tape.scene({
	caption: `docker stop ${CONTAINER} → liveness ping fails → banner flips to "${downState}"`,
	record: 3,
	hold: 6,
});

// ── Scene 3: bring it back + reconnect ──────────────────────────────
tape.log(`starting container ${CONTAINER}`);
await docker("start", CONTAINER);
// sshd needs a beat to be ready after start. Don't busy-poll the bridge —
// just give it a fixed grace period, then kick reconnect.
await tape.sleep(3500);

// Click Reconnect on the banner row if it offers one; fall back to the
// backend mutation so the scenario survives banner-styling tweaks.
const clicked = await tape.js<boolean>(
	`var r=document.querySelector('[data-testid=remote-connection-banner-row-${NAME}] button'); if(r){r.click(); return true;} return false;`,
);
if (clicked) {
	tape.log("clicked Reconnect button in banner");
} else {
	tape.log("no banner button; falling back to reconnect_remote_runtime");
	await tape.invoke("reconnect_remote_runtime", { name: NAME }, 60_000);
}

const recovered = await waitForState((s) => s === "connected", 60_000);
tape.assert("state_recovers", recovered === "connected", `state=${recovered}`);

// Banner should auto-disappear when state goes back to connected. Open the
// Remote Servers panel again so the row's "Connected" chip is on screen.
await tape.openSettings("remote-servers");
await tape.waitFor(`[data-testid=remote-server-row-${NAME}]`, 10_000);
await tape.sleep(800);

await tape.scene({
	caption: `Reconnect → SSH re-establishes → ${NAME} green again. No restart, no losing your work.`,
	hold: 6,
});

const passed = await tape.finish({ runtimeName: NAME, downState });
process.exit(passed ? 0 : 1);

File: src/scenarios/resilience.rs
//! Track C proof: when the remote host disappears (modeled by
//! `docker stop` of the container hosting helmor-server) Helmor's
//! liveness loop notices, flips the runtime to a non-`connected`
//! state, surfaces a top-of-shell banner, and a one-click Reconnect
//! re-establishes SSH once the host comes back.
//!
//! Rust port of `scenarios/resilience.ts`. Shells out to `docker stop`
//! and `docker start` for the chaos events; the rest is plain bridge
//! driving. Idempotent: if the runtime is already disconnected when
//! the scenario starts (the user kicked it before recording), it
//! reconnects first so the failure → recovery story stays meaningful.

use std::process::Command;
use std::time::Duration;

use anyhow::{anyhow, Result};
use serde::Deserialize;
use serde_json::{json, Value};
use tokio::time::Instant;

use crate::tape::{SceneSpec, Tape};

#[derive(Debug, Clone)]
pub struct Config {
    pub runtime_name: String,
    pub container: String,
}

impl Config {
    pub fn from_env() -> Self {
        Self {
            runtime_name: std::env::var("RUNTIME_NAME")
                .unwrap_or_else(|_| "docker-linux-arm64".into()),
            container: std::env::var("CONTAINER")
                .unwrap_or_else(|_| "helmor-test-linux-arm64".into()),
        }
    }
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct RuntimeEntry {
    pub name: String,
    pub state: Option<RuntimeState>,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct RuntimeState {
    #[serde(rename = "type")]
    pub kind: Option<String>,
}

fn docker(args: &[&str]) -> Result<String> {
    let out = Command::new("docker").args(args).output()?;
    if !out.status.success() {
        return Err(anyhow!(
            "docker {} → exit {}: {}",
            args.join(" "),
            out.status,
            String::from_utf8_lossy(&out.stderr).trim()
        ));
    }
    Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
}

/// Poll `list_remote_runtimes` until the named runtime's `state.type`
/// satisfies `predicate`, or `timeout` elapses. Returns the last
/// observed label (which may be "(missing)" or "(unknown)").
async fn wait_for_state<F>(
    tape: &mut Tape,
    runtime_name: &str,
    predicate: F,
    timeout: Duration,
) -> Result<String>
where
    F: Fn(&str) -> bool,
{
    let deadline = Instant::now() + timeout;
    let mut last = "(unknown)".to_string();
    while Instant::now() < deadline {
        let runtimes: Vec<RuntimeEntry> = tape.invoke("list_remote_runtimes", json!({})).await?;
        let entry = runtimes.iter().find(|r| r.name == runtime_name);
        last = entry
            .and_then(|r| r.state.as_ref())
            .and_then(|s| s.kind.clone())
            .unwrap_or_else(|| {
                if entry.is_none() {
                    "(missing)".into()
                } else {
                    "(unknown)".into()
                }
            });
        if predicate(&last) {
            return Ok(last);
        }
        tape.sleep(Duration::from_millis(500)).await;
    }
    Ok(last)
}

pub async fn run(tape: &mut Tape, config: &Config) -> Result<bool> {
    // 0. Make sure we start connected.
    let initial = wait_for_state(
        tape,
        &config.runtime_name,
        |s| s == "connected",
        Duration::from_secs(5),
    )
    .await?;
    if initial != "connected" {
        tape.log(&format!(
            "runtime {} state={initial}; reconnecting first",
            config.runtime_name
        ));
        tape.invoke::<Value>(
            "reconnect_remote_runtime",
            json!({"name": config.runtime_name}),
        )
        .await?;
        wait_for_state(
            tape,
            &config.runtime_name,
            |s| s == "connected",
            Duration::from_secs(30),
        )
        .await?;
    }

    tape.js::<Value>(r#"window.location.reload(); return "r";"#)
        .await?;
    tape.sleep(Duration::from_secs(6)).await;
    tape.open_settings("remote-servers").await?;
    let panel_opens = tape
        .wait_for("[role=dialog]", Duration::from_secs(5))
        .await?;
    tape.assert("panel_opens", panel_opens, "");
    let row_selector = format!("[data-testid=remote-server-row-{}]", config.runtime_name);
    let row_present = tape
        .wait_for(&row_selector, Duration::from_secs(10))
        .await?;
    tape.assert("row_present", row_present, "");

    // Baseline.
    tape.scene(
        SceneSpec::new(format!(
            "Baseline: {} is connected — helmor-server is alive on the container",
            config.runtime_name
        ))
        .hold_sec(4),
    )
    .await?;

    // Kill the host.
    tape.close_dialog().await?;
    tape.sleep(Duration::from_millis(400)).await;
    tape.log(&format!("stopping container {}", config.container));
    docker(&["stop", "-t", "1", &config.container])?;

    let down_state = wait_for_state(
        tape,
        &config.runtime_name,
        |s| s != "connected",
        Duration::from_secs(20),
    )
    .await?;
    tape.assert(
        "state_flips_offline",
        down_state != "connected",
        format!("state={down_state}"),
    );
    let banner_selector = format!(
        "[data-testid=remote-connection-banner-row-{}]",
        config.runtime_name
    );
    let banner_appears = tape
        .wait_for(&banner_selector, Duration::from_secs(10))
        .await?;
    tape.assert("banner_appears", banner_appears, "top-of-shell banner");
    tape.scene(
        SceneSpec::new(format!(
            "docker stop {} → liveness ping fails → banner flips to \"{down_state}\"",
            config.container
        ))
        .record_sec(3)
        .hold_sec(6),
    )
    .await?;

    // Bring it back + reconnect.
    tape.log(&format!("starting container {}", config.container));
    docker(&["start", &config.container])?;
    tape.sleep(Duration::from_millis(3500)).await;

    let reconnect_script = format!(
        r#"var r=document.querySelector({sel}); if(r){{r.click(); return true;}} return false;"#,
        sel = serde_json::to_string(&format!(
            "[data-testid=remote-connection-banner-row-{}] button",
            config.runtime_name
        ))?,
    );
    let clicked: bool = tape.js(&reconnect_script).await?;
    if clicked {
        tape.log("clicked Reconnect button in banner");
    } else {
        tape.log("no banner button; falling back to reconnect_remote_runtime");
        tape.invoke::<Value>(
            "reconnect_remote_runtime",
            json!({"name": config.runtime_name}),
        )
        .await?;
    }

    let recovered = wait_for_state(
        tape,
        &config.runtime_name,
        |s| s == "connected",
        Duration::from_secs(60),
    )
    .await?;
    tape.assert(
        "state_recovers",
        recovered == "connected",
        format!("state={recovered}"),
    );

    tape.open_settings("remote-servers").await?;
    tape.wait_for(&row_selector, Duration::from_secs(10))
        .await?;
    tape.sleep(Duration::from_millis(800)).await;

    tape.scene(
        SceneSpec::new(format!(
            "Reconnect → SSH re-establishes → {} green again. No restart, no losing your work.",
            config.runtime_name
        ))
        .hold_sec(6),
    )
    .await?;

    tape.finish(json!({
        "runtimeName": config.runtime_name,
        "downState": down_state,
    }))
    .await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn config_from_env_defaults() {
        unsafe {
            std::env::remove_var("RUNTIME_NAME");
            std::env::remove_var("CONTAINER");
        }
        let c = Config::from_env();
        assert_eq!(c.runtime_name, "docker-linux-arm64");
        assert_eq!(c.container, "helmor-test-linux-arm64");
    }

    #[test]
    fn docker_propagates_non_zero_exit_with_stderr() {
        let err = docker(&["nonexistent-subcommand-xyz"]).expect_err("must fail");
        let msg = err.to_string();
        // Errors should mention either non-zero exit OR the docker not-installed case.
        assert!(
            msg.contains("docker") || msg.contains("No such file"),
            "expected docker-related error: {msg}"
        );
    }
}
